lib/core/filtering/post_process_engine.rb

```diff
@@ -0,0 +1,37 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+# frozen_string_literal: true
+
+require 'core/filtering/post_process_result'
+require 'core/filtering/simple_rule'
+
+module Core
+  module Filtering
+    class PostProcessEngine
+      DEFAULT_DOMAIN = 'DEFAULT'
+
+      def initialize(job_description)
+        @ordered_rules = ordered_rules(job_description['filtering'])
+      end
+
+      def process(document)
+        @ordered_rules.each do |rule|
+          if rule.match?(document)
+            return PostProcessResult.new(document, rule)
```

Comment:
So what if the document is matched by more than one rule? Simple case:
* We have two rules: [ include :: color :: red], [ exclude :: name :: apple ]
* A document with { name: apple, color: red } comes in
* First matching rule is include, and we return a result immediately
* As a result, the document is indexed

Second case, and we reorder the rules:
* We have two rules: [ exclude :: name :: apple ], [ include :: color :: red]
* A document with { name: apple, color: red } comes in
* First matching rule is exclude, and we return a result immediately
* As a result, the document is not indexed

But both cases should have brought the same result right?

Reply:
Nope, ordering is very important and intentional for post-processing. This follows the behavior we've established and hardened in Crawl Rules. The UI lets you reorder rules. In the first example you provided, this would be good if you wanted to include red apples but not green apples, but all colors of pear (green AND red).

For pre-processing, order doesn't really matter, and it can't, if you're just ANDing all the clauses together. That discrepency is why we need to do our best effort to detect if simple rules cause conflicting pre-processing clauses. Again using your top example, if there were a third rule `[include :: name :: .*]`, we could detect that this conflicts with `[ exclude :: name :: apple ]`, and that we therefore shouldn't use _either_ for preprocessing.